core/consensus/grandpa/impl/voting_round_impl.cpp

```diff
@@ -790,29 +819,38 @@ namespace kagome::consensus::grandpa {
              proposal.getBlockHash(),
              proposal.id.toHex());
 
+    if (primary_vote_.has_value()) {
+      propagate = false;
+    }
+
     primary_vote_ = {{proposal.getBlockNumber(), proposal.getBlockHash()}};
+
+    if (propagate) {
+      sendProposal(convertToPrimaryPropose(proposal.getBlockInfo()));
+    }
   }
 
-  void VotingRoundImpl::onPrevote(const SignedMessage &prevote) {
+  bool VotingRoundImpl::onPrevote(const SignedMessage &prevote,
```

Comment:
I'd suggest an enum instead of a bool parameter to improve readability of the function calls. E. g.
`VotingRoundImpl::onPrevote(vote, ShouldPropagate::TRUE)`
instead of 
`VotingRoundImpl::onPrevote(vote, true)`

Reply:
I probably agree...